### PR TITLE
Adopt spring security

### PR DESCRIPTION
**Pull request by IvanBorislavovDimitrov**
LMCROSSITXSADEPLOY-1850

### CODE REVIEW

File: src/main/java/com/sap/cloudfoundry/client/facade/util/RestUtil.java [hunk 1]
```diff
@@ -26,13 +32,50 @@
...
+        URL authorizationUrl = getAuthorizationUrl(controllerUrl, webClient);
+        return new OAuthClient(authorizationUrl, webClient);
+    }
+
+    private URL getAuthorizationUrl(URL controllerUrl, WebClient webClient) {
+        Map<String, Object> infoMap = getInfoMap(controllerUrl, webClient);
+        return getAuthorizationUrl(infoMap);
+    }
+
+    // TODO: Refactor not to use v2
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/util/RestUtil.java:55]: v3 info is on root level
`v2/info` ->`/`

File: src/main/java/com/sap/cloudfoundry/client/facade/util/RestUtil.java [hunk 2]
```diff
@@ -26,13 +32,50 @@
...
+        try {
+            return objectMapper.readValue(infoResponse, new TypeReference<Map<String, Object>>() {
+            });
+        } catch (IOException e) {
+            throw new IllegalArgumentException("Error getting /v2/info from cloud controller.", e);
+        }
+    }
+
+    private URL getAuthorizationUrl(Map<String, Object> infoMap) {
+        String authorizationEndpoint = (String) infoMap.get("authorization_endpoint");
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/util/RestUtil.java:71]: The v3 information about the authorization url is changed.
See this commit https://github.com/cloudfoundry-incubator/multiapps-controller/commit/e96f7526ff9103a9c28dc868c31f9e9895a26b7e for more information

**Review by radito3 — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 3]
```diff
@@ -209,17 +209,32 @@
                 </exclusion>
             </exclusions>
         </dependency>
-        <!-- https://mvnrepository.com/artifact/org.springframework.security.oauth/spring-security-oauth2 -->
+        <!-- https://mvnrepository.com/artifact/org.springframework.security/spring-security-core -->
         <dependency>
-            <groupId>org.springframework.security.oauth</groupId>
-            <artifactId>spring-security-oauth2</artifactId>
-            <version>${spring-security-oauth2.version}</version>
-            <exclusions>
```
[the remark this answers is not in the document]
Comment (IvanBorislavovDimitrov) [reply pom.xml:217]: Just checked 10x, there aren't any

File: src/main/java/com/sap/cloudfoundry/client/facade/oauth2/OAuthClient.java [hunk 4]
```diff
@@ -48,67 +51,66 @@ public void clear() {
...
         return new OAuthTokenProvider(this);
     }
 
-    private AccessTokenRequest createAccessTokenRequest(String username, String password) {
-        return new DefaultAccessTokenRequest();
+    private boolean shouldRefreshToken() {
+        return credentials.isRefreshable() && token.getOAuth2AccessToken()
+                                                   .getExpiresAt()
+                                                   .isBefore(Instant.now()
+                                                                    .plus(120, ChronoUnit.SECONDS));
```
[the remark this answers is not in the document]
Comment (IvanBorislavovDimitrov) [reply src/main/java/com/sap/cloudfoundry/client/facade/oauth2/OAuthClient.java]: Not really

File: src/main/java/com/sap/cloudfoundry/client/facade/oauth2/OAuthClient.java [hunk 5]
```diff
@@ -48,67 +51,66 @@ public void clear() {
...
-        OAuth2AccessToken accessToken = getToken();
+    public String getAuthorizationHeaderValue() {
+        OAuth2AccessTokenWithAdditionalInfo accessToken = getToken();
         if (accessToken != null) {
-            return accessToken.getTokenType() + " " + accessToken.getValue();
+            return accessToken.getOAuth2AccessToken()
+                              .getTokenType()
+                              .getValue()
+                + " " + accessToken.getOAuth2AccessToken()
+                                   .getTokenValue();
```
[the remark this answers is not in the document]
Comment (IvanBorislavovDimitrov) [reply src/main/java/com/sap/cloudfoundry/client/facade/oauth2/OAuthClient.java]: Now it's a bit prettier

File: src/main/java/com/sap/cloudfoundry/client/facade/util/RestUtil.java (hunk 2 above)
Comment (IvanBorislavovDimitrov) [reply src/main/java/com/sap/cloudfoundry/client/facade/util/RestUtil.java]: Reused the code

File: src/main/java/com/sap/cloudfoundry/client/facade/oauth2/OAuthClient.java [hunk 6]
```diff
@@ -48,67 +51,62 @@ public void clear() {
...
+                            .retrieve()
+                            .bodyToFlux(Oauth2AccessTokenResponse.class)
+                            .blockFirst();
+        } catch (WebClientResponseException e) {
+            throw new ResponseStatusException(e.getStatusCode(), e.getMessage(), e);
+        }
+    }
 
-        return resource;
+    private TokenFactory getTokenFactory() {
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/oauth2/OAuthClient.java:109]: If this method is private and therefore it won't be overridden in tests, why not inline it?

File: src/main/java/com/sap/cloudfoundry/client/facade/oauth2/OAuthClientWithLoginHint.java [hunk 7]
```diff
@@ -56,10 +50,14 @@ private TokenProvider createTokenProvider() {
...
 
-    private OAuth2AccessToken getOrRefreshToken() {
+    private OAuth2AccessTokenWithAdditionalInfo getOrRefreshToken() {
         String token = getTokenProvider().getToken(connectionContext)
                                          .block();
-        return new DefaultOAuth2AccessToken(getTokenValue(token));
+        return getTokenFactory().createToken(getTokenValue(token));
+    }
+
+    private TokenFactory getTokenFactory() {
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/oauth2/OAuthClientWithLoginHint.java:59]: Same here

File: src/main/java/com/sap/cloudfoundry/client/facade/oauth2/TokenFactory.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    public static final String SCOPE_CC_WRITE = "cloud_controller.write";
+    public static final String SCOPE_CC_ADMIN = "cloud_controller.admin";
+    public static final String SCOPE_SCIM_USERIDS = "scim.userids";
+    public static final String SCOPE_PASSWORD_WRITE = "password.write";
+    public static final String SCOPE_OPENID = "openid";
+
+    // Token Body elements:
+    public static final String SCOPE = "scope";
+    public static final String EXP = "exp";
+    public static final String IAT = "iat";
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/oauth2/TokenFactory.java:31]: Rename these. They are hard to understand at first glance

File: src/main/java/com/sap/cloudfoundry/client/facade/oauth2/TokenFactory.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            return null;
+        }
+        return new OAuth2AccessTokenWithAdditionalInfo(createOAuth2AccessToken(tokenString, scope, expiresAt, instantiatedAt), tokenInfo);
+    }
+
+    private OAuth2AccessToken createOAuth2AccessToken(String tokenString, List<String> scope, Number expiresAt, Number instantiatedAt) {
+        try {
+            return new OAuth2AccessToken(OAuth2AccessToken.TokenType.BEARER,
+                                         tokenString,
+                                         new Date(instantiatedAt.longValue() * 1000).toInstant(),
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/oauth2/TokenFactory.java:56]: Why not use `Instant.ofEpochSeconds()` if the `instancedAt` and `expiresAt` variables are in seconds?

File: src/main/java/com/sap/cloudfoundry/client/facade/oauth2/TokenFactory.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+                                         new HashSet<>(scope));
+        } catch (IllegalArgumentException e) {
+            throw new ResponseStatusException(HttpStatus.UNAUTHORIZED, e.getMessage(), e);
+        }
+    }
+
+    private Map<String, Object> parseToken(String tokenString) {
+        String[] tokenParts = tokenString.split("\\.");
+        if (tokenParts.length != 3) {
+            // The token should have three parts (header, body and signature) separated by a dot. It doesn't, so we consider it as invalid.
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/oauth2/TokenFactory.java:67]: Extract the token string parts length into a named constant, so you can remove this comment

File: src/main/java/com/sap/cloudfoundry/client/facade/util/AuthorizationEndpointGetter.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    public String getAuthorizationEndpoint() {
+        String response = webClient.get()
+                                   .uri("/")
+                                   .retrieve()
+                                   .bodyToMono(String.class)
+                                   .block();
+        return getLoginHref(response);
+    }
```
Comment (radito3) [inline src/main/java/com/sap/cloudfoundry/client/facade/util/AuthorizationEndpointGetter.java:31]: Remove the duplicated code from the second method and call the first with an empty string for `controllerUrl`

**Review by radito3 — COMMENTED** [review]
Great that you've moved the `AuthorizationEndpointGetter` here, just don't forget to remove it from multiapps-controller

**Review by radito3 — APPROVED** [review]

**Review by radoslav-d — APPROVED** [review]

**Pull request merged**